src/core/umount.c

```diff
@@ -381,6 +383,112 @@ static bool nonunmountable_path(const char *path) {
+         * remount pid to exit. */
+        until = now(CLOCK_MONOTONIC) + DEFAULT_TIMEOUT_USEC;
+        for (;;) {
+                usec_t n;
+                struct timespec ts;
+
+                n = now(CLOCK_MONOTONIC);
+                if (n >= until)
+                        break;
+
+                r = sigtimedwait(&mask, NULL, timespec_store(&ts, until - n));
+                if (r < 0) {
+                        if (errno == EAGAIN) {
+                                /* Timed out, remount child is likely hung,
+                                 * but we can try. */
+                                if (waitpid(pid, &status, WNOHANG) == pid)
+                                        /* Success */
+                                        return (WEXITSTATUS(status) == 0 ? 0 : -WEXITSTATUS(status));
+                                else {
+                                        /* Fail, we should kill the hung
+                                         * process and return */
+                                        log_notice("Timed out, issuing SIGKILL to %d.", pid);
+
+                                        kill(pid, SIGKILL);
+                                        break;
+                                }
+                        } else if (errno == EINTR)
+                                /* Received a different signal - Retry */
+                                continue;
+                } else {
+                        if (waitpid(pid, &status, WNOHANG) == pid) {
+                                return (WEXITSTATUS(status) == 0 ? 0 : -WEXITSTATUS(status));
+                        } else
+                                /* Recieved a SIGCHLD, but not for the PID
+                                 * we were watching. We will retry */
+                                continue;
+                }
+                break;
+        }
+
+        return -ETIME;
+}
+
+static int remount_with_timeout(MountPoint *m, char *options, int *n_failed, bool log_error) {
+        pid_t pid;
+        int r;
+
+        /* Due to the possiblity of a remount operation hanging, we
+         * fork a child process and set a timeout. If the timeout
+         * lapses, the assumption is that that particular remount
+         * failed. */
+        pid = fork();
+        if (pid < 0)
+                return log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                log_info("Remounting '%s' read-only in with options '%s'.", m->path, options);
+
+                /* Start the mount operation here in the child */
+                r = mount(NULL, m->path, NULL, MS_REMOUNT|MS_RDONLY, options);
+                if (r < 0) {
+                        if (log_error)
+                                log_notice_errno(errno, "Failed to remount '%s' read-only: %m", m->path);
+                }
+                _exit(r < 0 ? EXIT_FAILURE : EXIT_SUCCESS);
+        }
+
+        return wait_for_child(pid);
```

Comment:
this is racy. we need to block SIGCHLD before fork()ing, and unblock after we waited correctly, so that no SIGCHLD is dispatched int the time between the fork() and the waitpid().

Consider using the nifty BLOCK_SIGNALS() macro for this